clustermesh-apiserver/etcdinit/root.go

```diff
@@ -0,0 +1,122 @@
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return InitEtcdLocal()
+		},
+	}
+	rootCmd.Flags().StringVar(&etcdDataDir, "data-dir", defaults.EtcdDataDir, "etcd data directory")
+	rootCmd.Flags().StringVar(&etcdClusterName, "cluster-name", defaults.EtcdClusterName, "etcd cluster name")
+	rootCmd.Flags().StringVar(&etcdInitialClusterToken, "initial-cluster-token", defaults.EtcdInitialClusterToken, "etcd initial cluster token")
+	rootCmd.Flags().BoolVar(&debug, "debug", false, "Debug log output")
+	return rootCmd
+}
+
+func InitEtcdLocal() error {
+	ctx, cancelFn := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancelFn()
+
+	if debug {
+		logging.SetLogLevelToDebug()
+		log.Debug("Debug logging enabled")
+	}
+
+	log := log.
+		WithField("etcdBinaryLocation", etcdBinaryLocation).
+		WithField("etcdDataDir", etcdDataDir).
+		WithField("etcdClusterName", etcdClusterName).
+		WithField("etcdInitialClusterToken", etcdInitialClusterToken)
+
+	// When the clustermesh-apiserver is launched, we create a new etcd. We don't support persistence of any kind
+	// here. So we just delete the contents of the data directory. It should be empty, but if the initialization failed
+	// we might re-run this as an init container.
+	log.Info("Deleting contents of data directory")
+	dir, err := os.ReadDir(etcdDataDir)
+	if err != nil {
+		return err
+	}
+	for _, d := range dir {
+		log.WithField("path", d.Name()).Debug("Removing file/directory in data dir")
+		os.RemoveAll(path.Join(etcdDataDir, d.Name()))
+	}
+
+	log.Info("Starting localhost-only etcd process")
+	// Specify the full path to the etcd binary to avoid any PATH search binary replacement nonsense
+	endpoint := "http://127.0.0.1:2379"
+	etcdCmd := exec.CommandContext(ctx, etcdBinaryLocation,
+		fmt.Sprintf("--data-dir=%s", etcdDataDir),
+		fmt.Sprintf("--name=%s", etcdClusterName),
+		// This probably won't work on an IPv6 only cluster
+		fmt.Sprintf("--listen-client-urls=%s", endpoint),
+		fmt.Sprintf("--advertise-client-urls=%s", endpoint),
+		fmt.Sprintf("--initial-cluster-token=%s", etcdInitialClusterToken),
+		"--initial-cluster-state=new",
+		"--auto-compaction-retention=1")
+
+	// Exec the etcd binary, which ultimately calls fork(2) under the hood. We're attached to STDIN and STDOUT for the
+	// new process. We don't wait on its completion, because it'll never complete of course.
+	err = etcdCmd.Start()
+	if err != nil {
+		return err
+	}
+	log = log.WithField("etcdPID", etcdCmd.Process.Pid)
+	log.Info("Local etcd server process started")
+
+	// With the etcd process launched, we need to construct a client to that process
+	config := client.Config{
+		Context:   ctx,
+		Endpoints: []string{endpoint},
+	}
+	c, err := client.New(config)
+	if err != nil {
+		return err
+	}
+
+	// Run the init commands
+	log.Info("Starting etcd init process")
+	einit.InitEtcd(ctx, c, etcdClusterName)
```

Comment:
Should we log/handle errors here?

Reply:
Yes, it's a mistake that they're not handled.